Use std::forward_list instead of std::map in irept by default [blocks: #3486]

tautschnig opened the pull request:
This reduces the size of an irept by 5 pointers (i.e., 40 bytes on
64-bit systems).

Also make typet constructor compatible with the STL shipped with GCC 5.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Use std::forward_list instead of std::map in irept by default

This reduces the size of an irept by 5 pointers (i.e., 40 bytes on
64-bit systems). On SV-COMP's ReachSafety-ECA with this change we can
perform 3819.81 symex_step calls per second, compared to 2752.28 calls
per second with the std::map configuration. The performance improvements
are spread across various `irept`-related operations.

kroening commented:
Can you split? The constructor seems obvious and trivial, whereas the change to list will require us to do a lot of consulting & benchmarking.

tautschnig commented:
@kroening The constructor change is now in #4460.

tautschnig commented:
Just for the record: on ReachSafety-ECA with this change we can perform 3819.81 `symex_step` calls per second, compared to 2752.28 calls per second with develop. I'll now try to figure out what exactly makes so much of a difference.

tautschnig commented:
It seems that just about any `irept`-related operation is cheaper (creating fresh `irept`s of course, but also `irept::operator==`).

tautschnig committed:
Add missing std::map include

irep.h provides std::map if, and only if, NAMED_SUB_IS_FORWARD_LIST is
not set.

Also cleanup include guards (pushing them to be the outermost
preprocessor directives) and indentation of includes.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d1a788b53e53f832dd16126cd4ad02bdd07973e2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106442057
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

martin-cs commented on src/util/irep.h:
```diff
@@ -18,7 +18,9 @@ Author: Daniel Kroening, kroening@kroening.com
 // #define HASH_CODE
-// #define NAMED_SUB_IS_FORWARD_LIST
+#ifndef NAMED_SUB_IS_FORWARD_LIST
+#  define NAMED_SUB_IS_FORWARD_LIST
```
I'm probably being thick here but why not just `#define` it?

tautschnig replied:
This ensures compatibility with any builds (such as our CI!) using `-DNAMED_SUB_IS_FORWARD_LIST` on the compiler command line.

martin-cs commented on src/memory-analyzer/gdb_api.h:
```diff
@@ -24,14 +27,13 @@ Author: Malte Mues <mail.mues@gmail.com>
+#  include <map>
 
-#include <util/exception_utils.h>
+#  include <util/exception_utils.h>
```
The indenting of includes seems inconsistent with the rest of the code.

tautschnig replied:
This is the preprocessor indentation used as of #4461 (which only applies to preprocessor directives).

kroening replied:
That one seems highly unusual -- can clang-format be made not to do that?

tautschnig replied:
In my opinion, this one is entirely appropriate, because of the `#if` that precedes it. But maybe I'm the only one thinking so and if that's the case we can of course avoid this by adding a `clang-format off`.

martin-cs replied:
How much of a diff would it produce if we indent vs. if we don't.  I think that changes to the coding style that are a departure from the existing style (regardless of whether they are "right" or "wrong") need some thought and justification.  To me 'consistent' is a key priority.

martin-cs reviewed (commented):
1. I think we're going to need some large scale benchmarking before we can make this change.

2. Might part of the performance benefit be fitting an `irept` into a single cache line?

3. My guess is that there is a clear bias on which of the entries of the map are most used.  Does it make sense to sort them / use a list that pivots them to the front?

tautschnig commented:
Although I responded with a 👍 I didn't actually provide meaningful comments to what @martin-cs said:

> I think we're going to need some large scale benchmarking before we can make this change.

Yes, all proposals or efforts most welcome.

> Might part of the performance benefit be fitting an `irept` into a single cache line?

Given that a typical cache line size today is 64B and that on 64-bit Linux systems we move from `sizeof(irept::dt) == 80` to `sizeof(irept::dt) == 40` you may very well be right!

> My guess is that there is a clear bias on which of the entries of the map are most used. Does it make sense to sort them / use a list that pivots them to the front?

We might as well want to measure the median size of the map: I'd suspect it would be 1 or 2 (expressions carry a type and maybe a source location).

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 0b0ee0591435dba607ff5fc6c11a887cd63c2540).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108706493
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

tautschnig commented:
@danpoe @smowton @romainbrenguier Would you be able to do some benchmarking of this change? I've also asked elsewhere, but it would be nice to have as much data as possible!

danpoe commented:
On a set of methods from Apache Tika, symex was overall 21% faster with the changes in this PR (20.8s vs 26.4s).

smowton commented:
On my large Webgoat run I find running time is practically identical, but the memory usage is down around 30%.

martin-cs commented:
> Although I responded with a 👍 I didn't actually provide meaningfu
> l comments to what @martin-cs said:

No worries -- I ignored your responses for a month! :-\


> > I think we're going to need some large scale benchmarking before we
> > can make this change.
> 
> Yes, all proposals or efforts most welcome.

Am a bit short on time but will try when I have some.


> > My guess is that there is a clear bias on which of the entries of
> > the map are most used. Does it make sense to sort them / use a list
> > that pivots them to the front?
> 
> We might as well want to measure the median size of the map: I'd
> suspect it would be 1 or 2 (expressions carry a type and maybe a
> source location). 

Yes; that would be the thing to do before thinking of any
optimisations.

romainbrenguier commented:
Here are the results on my benchmark:
![chart (3)](https://user-images.githubusercontent.com/11854194/57523263-6e667e00-731c-11e9-8cdb-adaf2f1a81d3.png)
overall forward lists were about 10% faster

martin-cs commented:
That's starting to look quite convincing isn't it?

tautschnig commented:
No known regressions (including external feedback), just needs final approval from at least one of @chrisr-diffblue @forejtv @peterschrammel @thk123.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 754dbfc39aee2fc15978f43abf08b630be7d678a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112051965
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: cfcc7e8e4070e9fde997676c742b14edd8c92777).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112097093
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 2b9849af73f150bab0ffd592b807aef7eea1feea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112365181
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
